File: app/src/test/java/com/mohsenoid/rickandmorty/data/db/DbTest.java
package com.mohsenoid.rickandmorty.data.db;

import android.app.Application;
import android.os.Build;

import androidx.test.core.app.ApplicationProvider;

import com.mohsenoid.rickandmorty.data.db.dto.DbCharacterModel;
import com.mohsenoid.rickandmorty.data.db.dto.DbEpisodeModel;
import com.mohsenoid.rickandmorty.data.exception.NoOfflineDataException;
import com.mohsenoid.rickandmorty.data.mapper.CharacterMapper;
import com.mohsenoid.rickandmorty.test.CharacterDataFactory;
import com.mohsenoid.rickandmorty.test.DataFactory;
import com.mohsenoid.rickandmorty.test.EpisodeDataFactory;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;

@Config(sdk = Build.VERSION_CODES.P)
@RunWith(RobolectricTestRunner.class)
public class DbTest {

    private Db db;

    @Before
    public void setUp() {
        Application application = ApplicationProvider.getApplicationContext();
        db = DbImpl.getInstance(application);
    }

    @After
    public void tearDown() {
        DbImpl.instance = null;
    }

    @Test
    public void testInsertEpisode() {
        // GIVEN
        DbEpisodeModel episode = EpisodeDataFactory.Db.makeDbEpisodeModel();

        // WHEN
        db.insertEpisode(episode);

        // THEN
        List<DbEpisodeModel> result = db.queryAllEpisodes(1);
        assertTrue(result.contains(episode));
    }

    @Test
    public void testUpdateEpisode() {
        // GIVEN
        int episodeId = DataFactory.randomInt();
        DbEpisodeModel oldEpisode = EpisodeDataFactory.Db.makeDbEpisodeModelWithId(episodeId);
        DbEpisodeModel updatedEpisode = EpisodeDataFactory.Db.makeDbEpisodeModelWithId(episodeId);

        // WHEN
        db.insertEpisode(oldEpisode);
        db.insertEpisode(updatedEpisode);

        // THEN
        List<DbEpisodeModel> result = db.queryAllEpisodes(1);
        assertFalse(result.contains(oldEpisode));
        assertTrue(result.contains(updatedEpisode));
    }

    @Test
    public void testInsertCharacter() {
        // GIVEN
        DbCharacterModel character = CharacterDataFactory.Db.makeDbCharacterModel();

        // WHEN
        db.insertCharacter(character);

        // THEN
        DbCharacterModel result = db.queryCharacter(character.getId());
        assertEquals(result, character);
    }

    @Test
    public void testUpdateCharacter() {
        // GIVEN
        int characterId = DataFactory.randomInt();
        DbCharacterModel oldCharacter = CharacterDataFactory.Db.makeDbCharacterModelWithId(characterId);
        DbCharacterModel updatedCharacter = CharacterDataFactory.Db.makeDbCharacterModelWithId(characterId);

        // WHEN
        db.insertCharacter(oldCharacter);
        db.insertCharacter(updatedCharacter);

        // THEN
        DbCharacterModel result = db.queryCharacter(characterId);
        assertNotEquals(result, oldCharacter);
        assertEquals(result, updatedCharacter);
    }

    @Test
    public void testQueryCharacter() {
        // GIVEN
        int characterId = DataFactory.randomInt();
        DbCharacterModel expected = CharacterDataFactory.Db.makeDbCharacterModelWithId(characterId);

        // WHEN
        db.insertCharacter(expected);

        // THEN
        DbCharacterModel actual = db.queryCharacter(characterId);
        assertEquals(expected, actual);
    }

    @Test
    public void testQueryAllCharacters() {
        // GIVEN
        DbCharacterModel expected = CharacterDataFactory.Db.makeDbCharacterModel();
        List<Integer> ids = Collections.singletonList(expected.getId());

        // WHEN
        db.insertCharacter(expected);

        // THEN
        List<DbCharacterModel> result = db.queryCharactersByIds(ids);
        assertTrue(result.contains(expected));
    }

    @Test
    public void testAliveCharacterNotKilledByUserIsAlive() {
        // GIVEN
        int characterId = DataFactory.randomInt();
        DbCharacterModel character = CharacterDataFactory.Db.makeDbCharacterModelWithIdAndStatus(characterId, CharacterMapper.CharacterDbMapper.ALIVE, true, false);

        // WHEN
        db.insertCharacter(character);

        // THEN
        DbCharacterModel result = db.queryCharacter(character.getId());
        assertEquals(result.getStatus(), CharacterMapper.CharacterDbMapper.ALIVE);
        assertTrue(result.getStatusAlive());
        assertFalse(result.getKilledByUser());
    }

    @Test
    public void testUserCanKillCharacter() throws NoOfflineDataException {
        // GIVEN
        int characterId = DataFactory.randomInt();
        DbCharacterModel character = CharacterDataFactory.Db.makeDbCharacterModelWithIdAndStatus(characterId, CharacterMapper.CharacterDbMapper.ALIVE, true, false);

        // WHEN
        db.insertCharacter(character);
        db.killCharacter(character.getId());

        // THEN
        DbCharacterModel result = db.queryCharacter(character.getId());
        assertEquals(result.getStatus(), CharacterMapper.CharacterDbMapper.ALIVE);
        assertTrue(result.getStatusAlive());
        assertTrue(result.getKilledByUser());
    }

    @Test
    public void testCharacterKilledByUserKeepsDeadAfterInsert() throws NoOfflineDataException {
        // GIVEN
        int characterId = DataFactory.randomInt();
        DbCharacterModel character = CharacterDataFactory.Db.makeDbCharacterModelWithIdAndStatus(characterId, CharacterMapper.CharacterDbMapper.ALIVE, true, false);

        // WHEN
        db.insertCharacter(character);
        db.killCharacter(character.getId());
        db.insertCharacter(character);

        // THEN
        DbCharacterModel result = db.queryCharacter(character.getId());
        assertEquals(result.getStatus(), CharacterMapper.CharacterDbMapper.ALIVE);
        assertTrue(result.getStatusAlive());
        assertTrue(result.getKilledByUser());
    }
}

File: app/src/test/kotlin/com/mohsenoid/rickandmorty/data/db/DbTest.kt
package com.mohsenoid.rickandmorty.data.db

import android.app.Application
import android.os.Build
import androidx.room.Room
import androidx.test.core.app.ApplicationProvider
import com.mohsenoid.rickandmorty.data.db.dao.DbCharacterDao
import com.mohsenoid.rickandmorty.data.db.dao.DbEpisodeDao
import com.mohsenoid.rickandmorty.data.db.entity.DbEntityCharacter
import com.mohsenoid.rickandmorty.data.db.entity.DbEntityEpisode
import com.mohsenoid.rickandmorty.data.mapper.CharacterDbMapper
import com.mohsenoid.rickandmorty.test.CharacterDataFactory
import com.mohsenoid.rickandmorty.test.DataFactory
import com.mohsenoid.rickandmorty.test.EpisodeDataFactory
import kotlinx.coroutines.runBlocking
import org.amshove.kluent.shouldBeFalse
import org.amshove.kluent.shouldBeTrue
import org.amshove.kluent.shouldContain
import org.amshove.kluent.shouldEqual
import org.amshove.kluent.shouldNotBeEmpty
import org.amshove.kluent.shouldNotBeNull
import org.amshove.kluent.shouldNotContain
import org.amshove.kluent.shouldNotEqual
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.koin.core.context.stopKoin
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import java.io.IOException

@Config(sdk = [Build.VERSION_CODES.P])
@RunWith(RobolectricTestRunner::class)
class DbTest {

    private lateinit var roomDb: RoomDb
    private lateinit var episodeDao: DbEpisodeDao
    private lateinit var characterDao: DbCharacterDao

    @Before
    fun setUp() {
        val application: Application = ApplicationProvider.getApplicationContext()

        roomDb = Room.inMemoryDatabaseBuilder(application, RoomDb::class.java).build()

        episodeDao = roomDb.episodeDao
        characterDao = roomDb.characterDao
    }

    @After
    @Throws(IOException::class)
    fun tearDown() {
        roomDb.close()
        stopKoin()
    }

    @Test
    fun `test insertEpisode can inserts a new Episode`() {
        runBlocking {
            // GIVEN
            val expectedEntityEpisode: DbEntityEpisode = EpisodeDataFactory.Db.makeEpisode()

            // WHEN
            episodeDao.insertEpisode(expectedEntityEpisode)

            // THEN
            val actualEntityEpisodes: List<DbEntityEpisode> = episodeDao.queryAllEpisodes()

            actualEntityEpisodes
                .shouldNotBeEmpty()
                .shouldContain(expectedEntityEpisode)
        }
    }

    @Test
    fun `test insertEpisode with same ID updates the old one`() {
        runBlocking {
            // GIVEN
            val episodeId: Int = DataFactory.randomInt()
            val oldEntityEpisode: DbEntityEpisode =
                EpisodeDataFactory.Db.makeEpisode(episodeId = episodeId)
            val updatedEntityEpisode: DbEntityEpisode =
                EpisodeDataFactory.Db.makeEpisode(episodeId = episodeId)

            // WHEN
            episodeDao.insertEpisode(oldEntityEpisode)
            episodeDao.insertEpisode(updatedEntityEpisode)

            // THEN
            val actualEntityEpisodes: List<DbEntityEpisode> = episodeDao.queryAllEpisodes()

            actualEntityEpisodes
                .shouldNotBeEmpty()
                .shouldContain(updatedEntityEpisode)
                .shouldNotContain(oldEntityEpisode)
        }
    }

    @Test
    fun `test insertCharacter can inserts a new Character`() {
        runBlocking {
            // GIVEN
            val expectedCharacter: DbEntityCharacter = CharacterDataFactory.Db.makeCharacter()

            // WHEN
            characterDao.insertOrUpdateCharacter(expectedCharacter)

            // THEN
            val actualCharacter: DbEntityCharacter? =
                characterDao.queryCharacter(expectedCharacter.id)

            actualCharacter
                .shouldNotBeNull()
                .shouldEqual(expectedCharacter)
        }
    }

    @Test
    fun `test insertCharacter with same ID updates the old one`() {
        runBlocking {
            // GIVEN
            val characterId: Int = DataFactory.randomInt()
            val oldCharacter: DbEntityCharacter =
                CharacterDataFactory.Db.makeCharacter(characterId = characterId)
            val updatedCharacter: DbEntityCharacter =
                CharacterDataFactory.Db.makeCharacter(characterId = characterId)

            // WHEN
            characterDao.insertOrUpdateCharacter(oldCharacter)
            characterDao.insertOrUpdateCharacter(updatedCharacter)

            // THEN
            val actualCharacter: DbEntityCharacter? = characterDao.queryCharacter(characterId)

            actualCharacter
                .shouldNotBeNull()
                .shouldEqual(updatedCharacter)
                .shouldNotEqual(oldCharacter)
        }
    }

    @Test
    fun `test queryCharacter returns the same Character details inserted`() {
        runBlocking {
            // GIVEN
            val characterId: Int = DataFactory.randomInt()
            val expectedCharacter: DbEntityCharacter =
                CharacterDataFactory.Db.makeCharacter(characterId = characterId)

            // WHEN
            characterDao.insertOrUpdateCharacter(expectedCharacter)

            // THEN
            val actualCharacter: DbEntityCharacter? = characterDao.queryCharacter(characterId)

            actualCharacter
                .shouldNotBeNull()
                .shouldEqual(expectedCharacter)
        }
    }

    @Test
    fun `test queryCharactersByIds returns the Characters asked for`() {
        runBlocking {
            // GIVEN
            val expected: DbEntityCharacter = CharacterDataFactory.Db.makeCharacter()
            val notExpected: DbEntityCharacter = CharacterDataFactory.Db.makeCharacter()
            val expectedCharacterIds: List<Int> = listOf(expected.id)

            // WHEN
            characterDao.insertOrUpdateCharacter(expected)
            characterDao.insertOrUpdateCharacter(notExpected)

            // THEN
            val characters: List<DbEntityCharacter> =
                characterDao.queryCharactersByIds(expectedCharacterIds)

            characters
                .shouldNotBeEmpty()
                .shouldContain(expected)
                .shouldNotContain(notExpected)
        }
    }

    @Test
    fun `test Character which status is Alive and is not KilledByUser isAlive`() {
        runBlocking {
            // GIVEN
            val expectedCharacterId: Int = DataFactory.randomInt()
            val expectedCharacter: DbEntityCharacter = CharacterDataFactory.Db.makeCharacter(
                characterId = expectedCharacterId,
                status = CharacterDbMapper.ALIVE,
                isAlive = true,
                isKilledByUser = false
            )

            // WHEN
            characterDao.insertOrUpdateCharacter(expectedCharacter)

            // THEN
            val actualCharacter: DbEntityCharacter? =
                characterDao.queryCharacter(expectedCharacterId)

            actualCharacter.shouldNotBeNull()
            actualCharacter.status shouldEqual CharacterDbMapper.ALIVE
            actualCharacter.statusAlive.shouldBeTrue()
            actualCharacter.killedByUser.shouldBeFalse()
        }
    }

    @Test
    fun `test user can kill a Character`() {
        runBlocking {
            // GIVEN
            val expectedCharacterId: Int = DataFactory.randomInt()
            val expectedCharacter: DbEntityCharacter = CharacterDataFactory.Db.makeCharacter(
                characterId = expectedCharacterId,
                status = CharacterDbMapper.ALIVE,
                isAlive = true,
                isKilledByUser = false
            )

            // WHEN
            characterDao.insertOrUpdateCharacter(expectedCharacter)
            characterDao.killCharacter(expectedCharacterId)

            // THEN
            val actualCharacter: DbEntityCharacter? =
                characterDao.queryCharacter(expectedCharacterId)

            actualCharacter.shouldNotBeNull()
            actualCharacter.status shouldEqual CharacterDbMapper.ALIVE
            actualCharacter.statusAlive.shouldBeTrue()
            actualCharacter.killedByUser.shouldBeTrue()
        }
    }

    @Test
    fun `test a Character which isKilledByUser keeps dead after update by insert`() {
        runBlocking {
            // GIVEN
            val expectedCharacterId: Int = DataFactory.randomInt()
            val expectedCharacter: DbEntityCharacter = CharacterDataFactory.Db.makeCharacter(
                characterId = expectedCharacterId,
                status = CharacterDbMapper.ALIVE,
                isAlive = true,
                isKilledByUser = false
            )

            // WHEN
            characterDao.insertOrUpdateCharacter(expectedCharacter)
            characterDao.killCharacter(expectedCharacterId)
            characterDao.insertOrUpdateCharacter(expectedCharacter)

            // THEN
            val actualCharacter: DbEntityCharacter? =
                characterDao.queryCharacter(expectedCharacterId)

            actualCharacter.shouldNotBeNull()
            actualCharacter.status shouldEqual CharacterDbMapper.ALIVE
            actualCharacter.statusAlive.shouldBeTrue()
            actualCharacter.killedByUser.shouldBeTrue()
        }
    }
}
